rfcs/20220221-3d_model_support.md

```diff
@@ -0,0 +1,181 @@
+  - [ ] https://github.com/WordPress/openverse-frontend/blob/bbccc36ed49186b369cc4a69693fbc6d1411f29a/src/components/VContentSwitcher/VContentSwitcherButton.vue
+  - [ ] https://github.com/WordPress/openverse-frontend/blob/7d74f033ddc4d35889d7c43756c36150beee4192/src/components/VContentSwitcher/VContentTypes.vue
+- [ ] Create a `model/index.vue` endpoint for 3D models in the [pages directory](https://github.com/WordPress/openverse-frontend/blob/037f2efb768fdbdcff4b875e7f7c271ff50df59f/src/pages)
+- [ ] Add 3D model sources to the [legacy sources lookup](https://github.com/WordPress/openverse-frontend/blob/268532a05004b72fb91b77a875f3a3c1d7fb0599/src/utils/get-legacy-source-url.js#L1)
+- [ ] Add 3D models to the [media constants](https://github.com/WordPress/openverse-frontend/blob/4aeb9fd0d9f893f2df1752d6bb1e52351b8c3d35/src/constants/media.js#L13) file, making sure it is set to the `ADDITIONAL` status.
+
+### 2. Setup the data for 3D models
+
+- [ ] Create a feature flag for 3D model support. All 3D model features will be behind this.
+- [ ] Generate mock/sample api response data for 3D models.
+- [ ] Create a new media service for Models in the [data directory](https://github.com/WordPress/openverse-frontend/blob/fe24ec9bf6a4d3ad429901cef53744dccd93ed5d/src/data)
+- [ ] Update the [media constants](https://github.com/WordPress/openverse-frontend/blob/4aeb9fd0d9f893f2df1752d6bb1e52351b8c3d35/src/constants/media.js#L1) file with new constants for models, and under the `BETA` flag.
+- [ ] Update the [media store](https://github.com/WordPress/openverse-frontend/blob/f9f549beace4f76480c8cbf1e91d8940a0aeef2b/src/store/media.js#L347) with 3D model support.
+
+### 3. Implement new 3D model UI 
+
+#### Prerequisites
+
+Design work for the new components must be completed, or at least in a good enough place that we decide work can begin.
+
+#### Todos
+
+- [ ] Add 3D models to the All Results view
+  - [ ] Create a `VModelCell.vue` component in the [`VAllResultsGrid` directory](https://github.com/WordPress/openverse-frontend/blob/f5b7518c3942a2ede49d5c7fef2831d30e9867a9/src/components/VAllResultsGrid)
+  - [ ] Add the logic to display models in the [`VAllResultsGrid.vue`](https://github.com/WordPress/openverse-frontend/blob/8214217f84099a83b9767f6a5c73fdbdf278bc95/src/components/VAllResultsGrid/VAllResultsGrid.vue#L1) component.
+- [ ] Create a new single model route `pages/model/_id.vue` for viewing a 3D model in detail.
+	- [ ] Render the model
+	- [ ] Render the model info
+	- [ ] Render the attribution information
+	- [ ] Render the related models
+- [ ] Create a new `VThreeDModel.vue` component that supports displaying 3D models of various file formats
+
+## Open Questions
+
+- What model formats do we want/need to support?
+- What features do people need in a 3D model search engine? What functionality does our model viewer need to have?
+- Should we use SketchFab's viewer for SketchFab models?
```

Comment:
Copying my remarks and @zackkrida's additional background information from the issue:

I would really really like to avoid SketchFab's viewer if possible.

1. It is not free software
2. It has much [looser/worse privacy guarantees](https://help.sketchfab.com/hc/en-us/articles/203059088-Compatibility#troubleshooting-scripts) than we've tried to achieve with Openverse so far

On the other hand, their viewer is extremely convenient and works very well.

In a brief search I found this [js-3d-model-viewer](https://github.com/cgwire/js-3d-model-viewer) library that is MIT licensed. It's not great, but I wonder if we could help improve it and add some of the features we might want (like UI buttons for controlling zoom when a mousewheel isn't available, recentering, snapping to specific axis, etc).

Currently it only support OBJ files though and has poor texture support (the latter of which is something we could probably spend some time contributing against). These might be unacceptable limitations though.

I guess it depends on how much upfront work we're willing to do for this or if it's something we'd rather just get out the door and be able to move on to video.

**From Zack:**

Regarding viewers, I chatted a bit with @antpb today who works on wp core media, but also around open source projects tackling standardizing 3D content performance, and might have some recommendations for us here smile

Reply:
Hey there! I have been pretty heavy involved in 3D web development over the last several years and have been particularly focused on Three.js and glTF as my go-to stack for rendering 3D in a browser. I have a plugin that was released a couple weeks ago that officially supports three.js, WebXR and glTF files in the WordPress block editor. You can try the plugin here: https://wordpress.org/plugins/three-object-viewer/

I feel pretty strongly that glTF, fbx, Usdz, and obj will be the most asked for supports. React three fiber is a library that supports gltf and fbx out of the box and I have an example of an obj loader relative to three.js that I would be willing to contribute. React Three Fiber is what I'm using for the render and it is very plug and play with Gutenberg, so implementations of 3D seem like they would be more consistent with a library such as this. Hopefully my plugin will help give inspiration around how people want to use 3D files on their sites. I have many more features like animations and interactions in the pipeline!

Separately from the plugin above, I am part of a group working toward standards around 3D content on the web. The group is called Open Metaverse Interoperability Group: https://omigroup.org and it has been very successful in building a community of leaders in the open metaverse pursuit. All code within the group adheres to the w3c requirements and is licensed MIT. The group is looking for wider support of these extensions as time goes on so Openverse seems a really great pairing! 

( More on the w3c side of OMI here: https://www.w3.org/community/metaverse-interop/  )

The problem currently that the group is solving is around a lack of standards with glTF when it comes to more rich interactions (such as audio, interactive elements, etc). Every engine or renderer requires a unique implementation to achieve these features which means the content (3D object) usually only works for one. The group is solving part of this problem through glTF extensions which allow for interoperability across many environments. Currently the team has developed an audio emitter spec that has resulted in a node package that allows for adding a single parser to support both spatial and global audio in a three.js scene. These standards are not specific to three.js which is what makes the support of the gltf-extensions interoperable across many engines. An example of an extension spec is here: https://github.com/omigroup/gltf-extensions

Also here is the plug and play three.js package that enables files with these extensions to work like magic: https://www.npmjs.com/package/three-omi

A live example of this audio emitter can be found here: https://3ov.xyz/omi-audio-demo/

The above package was implemented in my plugin and you can explore the different features at my plugin's website here: https://3ov.xyz/

Hopefully this info helps. I realized I dumped a bunch of links but I'm always happy to talk about this topic as I have given it _a lot_ of thought. :D

Reply:
Despite the fact that Openverse's frontend is written in Vue.js, it _could still be_ entirely appropriate to use react-three-fiber regardless. Thanks for your thoughts!

Reply:
I don't think we should load react for this, but it does seem like there are some Vue equivalents: https://github.com/troisjs/trois

It appears to be more or less the same thing as react-three-fiber. It would need some work to get [this issue closed](https://github.com/troisjs/trois/issues/70) before we could use it though.

Reply:
I wouldn't be surprised if some renderers are even larger than react and react-dom, so while it's unlikely it's not completely off the table for me, but Trois looks amazing! Well maintained and it would be fun to contribute anything we find missing.

Reply:
That's true. I guess either way with this work we'll need to be careful to make sure we're lazy-loading/bundle splitting the rendering libraries and tools anyway; they are indeed quite heavy.